Add: credential login flow

ProgramStuff opened the pull request:
This includes updates to google login flow, to create a new user and link their social account. Credential provider has been added and interacts with db

ProgramStuff committed:
Add: credential login flow

This includes updates to google login flow, to create a new user and link their social account. Credential provider has been added and interacts with db

ProgramStuff committed:
Fix: token structure and credential login

Fix error where unauthorized when signing in using credentials.

Modify token and session structure to return user data

ProgramStuff committed:
Update: profile page

Update page to fetch session and render the users name

brueningf commented on frontend/src/auth.ts:
```diff
@@ -1,8 +1,25 @@
+import { object, string } from "zod";
 
 import GoogleProvider from "next-auth/providers/google";
 
```
Prisma cannot be instantiated on the frontend. move all db logic to the backend

brueningf commented on frontend/src/auth.ts:
```diff
@@ -1,8 +1,25 @@
 
+const prisma = new PrismaClient();
+
+const signInSchema = object({
```
Move validation to its own file

brueningf commented on frontend/src/auth.ts:
```diff
@@ -11,61 +28,126 @@ export const { handlers, auth, signIn, signOut } = NextAuth({
     ],
     session: { strategy: "jwt" },
     callbacks: {
         async signIn({ user, account, profile, email, credentials }) {
```
I think all logic that is about authorization should be in authorization callbacks. like for credentials, there is a callback for google credentials.

brueningf replied:
session should contain the same properties in either case

brueningf commented on frontend/src/auth.ts:
```diff
@@ -11,61 +28,126 @@ export const { handlers, auth, signIn, signOut } = NextAuth({
             }
+            return false;
+        },
+        async redirect({ url, baseUrl }) {
```
We should redirect back if the user was trying to enroll in a course, so check if url was a course url.

brueningf commented on frontend/src/app/user/profile/page.tsx:
Not sure if we need client code here, try to fetch session on the server, and render instead of rendering on the client. Also a unauthenticated users should not be able to reach this route, use middleware.ts to protect the route.
https://nextjs.org/docs/app/building-your-application/routing/middleware

brueningf reviewed (changes requested):
Move db logic to the backend, avoid using client libraries like react unless necessary, and move validation to its own file. To have a solid session object please update the session interface, and handle authorization in provider s callback. clean up general callbacks of auth.ts

ProgramStuff committed:
Update user signin endpoint

Moved all db logic for user signin and updated request to include password and account

ProgramStuff committed:
Update providers

Update google provider for the expected profile structure.

Moved db logic to the backend and validation to separate file

Added check for request coming from /courses

Updated JWT and Session interface

ProgramStuff committed:
Merge branch 'staging' into staging

brueningf commented on frontend/src/auth.ts:
```diff
@@ -76,12 +154,18 @@ export const { handlers, auth, signIn, signOut } = NextAuth({
     interface Session {
         accessToken?: string;
-        user?: Prisma.UserGetPayload<{}>;
+
```
shouldn't this be inside user?

brueningf commented on backend/src/routes/users.ts:
```diff
@@ -6,8 +6,8 @@ const router = Router();
 const ctx = createContext();
 
 // POST /signin
```
why remove types? we are using typescript be explicit about types.

brueningf commented on backend/src/routes/users.ts:
There is an issue here. This endpoint should handle both signups, and sign-ins from both credentials, and other providers. Right now there is no way to sign up. A password may not be needed since other providers use account tokens.

ProgramStuff committed:
Update middleware

Add protected user routes and signin route

ProgramStuff committed:
Merge branch 'staging' of https://github.com/ProgramStuff/V53-tier3-team-35 into staging

ProgramStuff committed:
Fix session issues/changes to schema

Login with credentials will now return a valid response.

Use server method to retrieve session

Add signout to nav bar

Add password field to user table

ProgramStuff committed:
Update schema/factory

Add password to schema
Update user factory with  password

ProgramStuff committed:
Add custom signin page

Modify auth to include name for credential and use custom signin page, automatically updating when new provers added.

Add hashing and comparing password
Include types

ProgramStuff committed:
Create signin page

Custom signin page for providers,  registration section for credentials

ProgramStuff committed:
Fix test failure

If undefined password will be null

ProgramStuff committed:
Fix missing migration file with user password

ProgramStuff committed:
Merge branch 'staging' into staging

ProgramStuff committed:
Add bcrypt to package.json

ProgramStuff committed:
Merge branch 'staging' of https://github.com/ProgramStuff/V53-tier3-team-35 into staging

ProgramStuff committed:
Include zod dependency

brueningf commented on backend/prisma/migrations/20250129120205_add_user_password/migration.sql:
I'll remove migrations for now

brueningf committed:
Delete backend/prisma/migrations/20250129120205_add_user_password/migration.sql

brueningf commented on frontend/src/app/page.tsx:
remove this. You may add small client components, but try not to make a whole page render on client. at least isolate to the page you want to render on client. It beats the purpose of using the app router, or even Next.js.

brueningf commented on frontend/src/app/signin/page.tsx:
```diff
@@ -0,0 +1,126 @@
+
+    return (
+        <div className="flex flex-col gap-2">
+            <br />
```
Do not use break line in templates.

brueningf commented:
Good job @ProgramStuff 
Let's fix that deploy before merging. You can try and build on your local env to see errors. I'll leave below the deployment error log that I saw on Netlify.


```
Type error: Type '{ searchParams: { callbackUrl: string | undefined; }; }' does not satisfy the constraint 'PageProps'.
3:23:53 PM:   Types of property 'searchParams' are incompatible.
3:23:53 PM:     Type '{ callbackUrl: string | undefined; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]
3:23:53 PM: Static worker exited with code: 1 and signal: null
```

ProgramStuff committed:
Remove custom signin page

ProgramStuff committed:
Merge branch 'staging' of https://github.com/ProgramStuff/V53-tier3-team-35 into staging

brueningf reviewed (approved):
Good work
